kanidmd/src/lib/idm/oauth2.rs

```diff
@@ -0,0 +1,1025 @@
+#[serde(rename_all = "lowercase")]
+pub enum Oauth2Error {
+    // Non-standard
+    AuthenticationRequired,
+    // Standard
+    InvalidRequest,
+    UnauthorizedClient,
+    AccessDenied,
+    UnsupportedResponseType,
+    InvalidScope,
+    ServerError(OperationError),
+    TemporarilyUnavailable,
+}
+
+impl std::fmt::Display for Oauth2Error {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str(match self {
+            Oauth2Error::AuthenticationRequired => "authentication_required",
+            Oauth2Error::InvalidRequest => "invalid_request",
+            Oauth2Error::UnauthorizedClient => "unauthorized_client",
+            Oauth2Error::AccessDenied => "access_denied",
+            Oauth2Error::UnsupportedResponseType => "unsupported_response_type",
+            Oauth2Error::InvalidScope => "invalid_scope",
+            Oauth2Error::ServerError(_) => "server_error",
+            Oauth2Error::TemporarilyUnavailable => "temporarily_unavailable",
+        })
+    }
+}
+
+// == internal state formats that we encrypt and send.
+
+#[derive(Serialize, Deserialize, Debug, PartialEq)]
+struct ConsentToken {
+    pub client_id: String,
+    // Must match the session id of the Uat,
+    pub session_id: Uuid,
+    // So we can ensure that we really match the same uat to prevent confusions.
+    pub ident_id: IdentityId,
+    // CSRF
+    pub state: Base64UrlSafeData,
+    // The S256 code challenge.
+    pub code_challenge: Base64UrlSafeData,
+    // Where the RS wants us to go back to.
+    pub redirect_uri: Url,
+}
+
+// consent token?
+
+#[derive(Serialize, Deserialize, Debug)]
+struct TokenExchangeCode {
+    // We don't need the client_id here, because it's signed with an RS specific
+    // key which gives us the assurance that it's the correct combination.
+    pub uat: UserAuthToken,
+    // The S256 code challenge.
+    pub code_challenge: Base64UrlSafeData,
+    // The original redirect uri
+    pub redirect_uri: Url,
+}
+
+// consentPermitResponse
+
+#[derive(Debug)]
+pub struct AuthorisePermitSuccess {
+    // Where the RS wants us to go back to.
+    pub redirect_uri: Url,
+    // The CSRF as a string
+    pub state: Base64UrlSafeData,
+    // The exchange code as a String
+    pub code: String,
+}
+
+// The cache structure
+
+#[derive(Clone)]
+pub struct Oauth2RSBasic {
+    name: String,
+    uuid: Uuid,
+    origin: Origin,
+    authz_secret: String,
+    token_fernet: Fernet,
+}
+
+impl std::fmt::Debug for Oauth2RSBasic {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        f.debug_struct("Oauth2RSBasic")
+            .field("name", &self.name)
+            .field("uuid", &self.uuid)
+            .field("origin", &self.origin)
+            .finish()
+    }
+}
+
+#[derive(Debug, Clone)]
+pub enum Oauth2RS {
+    Basic(Oauth2RSBasic),
+}
+
+#[derive(Clone)]
+struct Oauth2RSInner {
+    fernet: Fernet,
+    rs_set: HashMap<String, Oauth2RS>,
+}
```

Comment:
What does "Inner" mean? So it is a cache that stores the metadata about all the resource servers?

Reply:
The "Inner" is because this structure is stored in a concurrently readable cell type (see Oauth2ResourceServers). So the cowcell provides concurrency and transactions, and the "inner" referes to the stored data within (the inner component) of the concurrent cell.